src/data/battler-tags.ts

```diff
@@ -3251,66 +3256,51 @@ export class TauntTag extends MoveRestrictionBattlerTag {
-      1,
-      MoveId.IMPRISON,
-      sourceId,
-    );
+export class ImprisoningTag extends BattlerTag implements RestrictingBattlerTag {
+  constructor() {
+    super(BattlerTagType.IMPRISONING, BattlerTagLapseType.CUSTOM, 1);
   }
 
-  /**
-   * Checks if the source of Imprison is still active
-   * @override
-   * @param pokemon The pokemon this tag is attached to
-   * @returns `true` if the source is still active
-   */
-  public override lapse(pokemon: Pokemon, lapseType: BattlerTagLapseType): boolean {
-    const source = this.getSourcePokemon();
-    if (source) {
-      if (lapseType === BattlerTagLapseType.PRE_MOVE) {
-        return super.lapse(pokemon, lapseType) && source.isActive(true);
-      } else {
-        return source.isActive(true);
-      }
-    }
-    return false;
+  override onAdd(pokemon: Pokemon): void {
+    globalScene.queueMessage(
+      i18next.t("battlerTags:imprisonOnAdd", { pokemonNameWithAffix: getPokemonNameWithAffix(pokemon) }),
+    );
   }
 
   /**
-   * Checks if the source of the tag has the parameter move in its moveset and that the source is still active
-   * @override
-   * @param moveId the move under investigation
-   * @returns `false` if either condition is not met
-   */
-  public override isMoveRestricted(moveId: MoveId, _user: Pokemon): boolean {
-    const source = this.getSourcePokemon();
-    if (source) {
-      const sourceMoveset = source.getMoveset().map((m) => m.moveId);
-      return sourceMoveset?.includes(moveId) && source.isActive(true);
+   * Prevents opposing Pokemon from using moves that match with
+   * the tag owner's moveset.
+   * @param pokemon The {@linkcode Pokemon} with this tag
+   * @param simulated If `true`, suppresses the message triggered by an interruption
+   * @param actingPokemon The {@linkcode Pokemon} attempting to select or use a move
+   * @param moveId The {@linkcode MoveId} for the move being selected or used
+   * @returns `true` if the given move is disabled by this tag
+   */
+  override apply(pokemon: Pokemon, simulated: boolean, actingPokemon: Pokemon, moveId: MoveId): boolean {
+    if (pokemon.getMoveset().some((mv) => mv.moveId === moveId)) {
+      if (!simulated) {
+        globalScene.queueMessage(this.interruptedText(actingPokemon, moveId));
+      }
+      return true;
     }
     return false;
   }
 
-  override selectionDeniedText(pokemon: Pokemon, moveId: MoveId): string {
+  public interruptedText(pokemon: Pokemon, moveId: MoveId): string {
     return i18next.t("battle:moveDisabledImprison", {
       pokemonNameWithAffix: getPokemonNameWithAffix(pokemon),
       moveName: allMoves[moveId].name,
     });
   }
 
-  override interruptedText(pokemon: Pokemon, moveId: MoveId): string {
-    return i18next.t("battle:moveDisabledImprison", {
-      pokemonNameWithAffix: getPokemonNameWithAffix(pokemon),
-      moveName: allMoves[moveId].name,
-    });
+  public selectionDeniedText(pokemon: Pokemon, moveId: MoveId): string {
```

Comment:
hmm, guess this also would need to be renamed then. Wonder how many similar methods there are, maybe it's not for this PR if there are a lot
```suggestion
  public getSelectionDeniedText(pokemon: Pokemon, moveId: MoveId): string {
```